{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TensorFlow version: 1.12.0\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "print (\"TensorFlow version: \" + tf.__version__)\n",
    "\n",
    "from keras.preprocessing.text import Tokenizer\n",
    "from keras.preprocessing.sequence import pad_sequences\n",
    "from keras.utils.np_utils import to_categorical\n",
    "from keras.layers import Embedding\n",
    "from keras.layers import Dense, Input, Flatten\n",
    "from keras.layers import Conv1D, MaxPooling1D, Embedding, Dropout, LSTM, GRU, Bidirectional\n",
    "from keras.models import Model\n",
    "from keras.callbacks import ModelCheckpoint\n",
    "import matplotlib.pyplot as plt\n",
    "plt.switch_backend('agg')\n",
    "from keras import backend as K\n",
    "from keras.engine.topology import Layer, InputSpec\n",
    "from keras import initializers\n",
    "%matplotlib inline\n",
    "\n",
    "# import data processing and other libraries\n",
    "import csv\n",
    "import sys\n",
    "import requests\n",
    "import os\n",
    "import re\n",
    "import pprint\n",
    "import pymongo\n",
    "import traceback\n",
    "from time import sleep\n",
    "import requests\n",
    "import pandas as pd\n",
    "import io\n",
    "from IPython.display import display\n",
    "import time\n",
    "import numpy as np\n",
    "from bs4 import BeautifulSoup\n",
    "import string\n",
    "import random\n",
    "from urllib.parse import urlparse, urljoin\n",
    "from collections import defaultdict\n",
    "from collections import OrderedDict\n",
    "import collections\n",
    "import pickle\n",
    "from collections import defaultdict\n",
    "\n",
    "from boilerpipe.extract import Extractor\n",
    "\n",
    "import nltk\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.stem.snowball import SnowballStemmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/10\n",
      "1000/1000 [==============================] - 0s 312us/step - loss: 0.7106 - acc: 0.4730\n",
      "Epoch 2/10\n",
      "1000/1000 [==============================] - 0s 36us/step - loss: 0.6985 - acc: 0.5130\n",
      "Epoch 3/10\n",
      "1000/1000 [==============================] - 0s 39us/step - loss: 0.6928 - acc: 0.5130\n",
      "Epoch 4/10\n",
      "1000/1000 [==============================] - 0s 46us/step - loss: 0.6885 - acc: 0.5350\n",
      "Epoch 5/10\n",
      "1000/1000 [==============================] - 0s 38us/step - loss: 0.6887 - acc: 0.5240\n",
      "Epoch 6/10\n",
      "1000/1000 [==============================] - 0s 56us/step - loss: 0.6838 - acc: 0.5650\n",
      "Epoch 7/10\n",
      "1000/1000 [==============================] - 0s 70us/step - loss: 0.6800 - acc: 0.5780\n",
      "Epoch 8/10\n",
      "1000/1000 [==============================] - 0s 46us/step - loss: 0.6789 - acc: 0.5700\n",
      "Epoch 9/10\n",
      "1000/1000 [==============================] - 0s 47us/step - loss: 0.6767 - acc: 0.5810\n",
      "Epoch 10/10\n",
      "1000/1000 [==============================] - 0s 38us/step - loss: 0.6762 - acc: 0.5810\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<keras.callbacks.History at 0x1278b1a20>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For a single-input model with 2 classes (binary classification):\n",
    "\n",
    "model = Sequential()\n",
    "model.add(Dense(32, activation='relu', input_dim=100))\n",
    "model.add(Dense(1, activation='sigmoid'))\n",
    "model.compile(optimizer='rmsprop',\n",
    "              loss='binary_crossentropy',\n",
    "              metrics=['accuracy'])\n",
    "\n",
    "# Generate dummy data\n",
    "import numpy as np\n",
    "data = np.random.random((1000, 100))\n",
    "labels = np.random.randint(2, size=(1000, 1))\n",
    "\n",
    "# Train the model, iterating on the data in batches of 32 samples\n",
    "model.fit(data, labels, epochs=10, batch_size=32)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load labeled data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "MONGODB_DB = \"FirmDB_20181226\"\n",
    "MONGODB_COLLECTION = \"pages_ABOUT2\"\n",
    "CONNECTION_STRING = \"mongodb://localhost\"\n",
    "\n",
    "client = pymongo.MongoClient(CONNECTION_STRING)\n",
    "db = client[MONGODB_DB]\n",
    "col = db[MONGODB_COLLECTION]\n",
    "\n",
    "ABOUT_DIR = '/Users/sarora/dev/EAGER/data/orgs/about/'\n",
    "DATA_DIR = '/Users/sarora/dev/EAGER/data/orgs/parsed_page_output/'\n",
    "GLOVE_DIR = '/Users/sarora/Documents/web of innovation/data/'\n",
    "\n",
    "TRAINING_PERCENT = .10\n",
    "PHRASE_LENGTH = 60\n",
    "MIN_PARA_LEN = 5\n",
    "pp = pprint.PrettyPrinter()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_domain (url):\n",
    "    o = urlparse(url.lower())\n",
    "    domain = o.netloc.strip('www.')\n",
    "    return domain\n",
    "\n",
    "# output urls for labeling of training data\n",
    "results = col.find({},{\"url\": 1, \"firm_name\": 1})\n",
    "df = pd.DataFrame(columns = ('firm_name', 'url', 'label'))\n",
    "domain_count = defaultdict(lambda:0,{})\n",
    "for i in range(results.count()):\n",
    "    result = results.next()\n",
    "    url = result['url'][0]\n",
    "    domain_count[get_domain(url)] += 1\n",
    "    firm_name = result['firm_name'][0] if 'firm_name' in result else ''\n",
    "    df.loc[i] = [firm_name, url, '']\n",
    "\n",
    "firm_names = list(set(df['firm_name']))\n",
    "df['gid'] = df.groupby(['firm_name']).ngroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['about_lbl', 'mgmt_lbl', 'partners_lbl', 'ip_lbl', 'about_agg_lbl', 'gid', 'pages_in_domain', 'is_sole_page']\n"
     ]
    }
   ],
   "source": [
    "# read back labeled data (note that about, management/team and partners, are dichotomous)\n",
    "df_about_labeled = pd.read_csv(ABOUT_DIR + 'about_pages_labeled_v3.csv')\n",
    "df_about_labeled = df_about_labeled.fillna(0)\n",
    "\n",
    "# count pages per domain\n",
    "for index, row in df_about_labeled.iterrows():\n",
    "    pages_in_domain = domain_count[get_domain(row['url'])]\n",
    "    df_about_labeled.loc[index,'pages_in_domain'] = pages_in_domain\n",
    "    is_sole_page = 0 if pages_in_domain > 1 else 1\n",
    "    df_about_labeled.loc[index,'is_sole_page'] = is_sole_page\n",
    "    \n",
    "labeled_urls = list(df_about_labeled['url']) # for training models on labeled urls below = \n",
    "about_labels = list(df_about_labeled['about_lbl'])\n",
    "mgmt_labels = list(df_about_labeled['mgmt_lbl'])\n",
    "partner_labels = list(df_about_labeled['partners_lbl'])\n",
    "ip_labels = list(df_about_labeled['ip_lbl'])\n",
    "df_about_labeled = df_about_labeled.set_index(['firm_name', 'url'])\n",
    "print (df_about_labeled.columns.tolist())\n",
    "\n",
    "# final test set is the rows of the original data frame without the urls in df_about_labeled "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['https://www.3m.com/', 'https://www.3m.com/3M/en_US/company-us/site-map/', 'https://www.3m.com/3M/en_US/company-us/about-3m/', 'https://www.3m.com/3M/en_US/home-improvement-us/', 'https://www.3m.com/3M/en_US/company-us/SDS-search/', 'https://www.3m.com/3M/en_US/company-us/patent/', 'https://www.3m.com/3M/en_US/company-us/help-center/', 'http://acaciaresearch.com/corporate-governance/', 'http://acaciaresearch.com/history/', 'http://acaciaresearch.com/corporate-profile/']\n",
      "[0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>about_lbl</th>\n",
       "      <th>mgmt_lbl</th>\n",
       "      <th>partners_lbl</th>\n",
       "      <th>ip_lbl</th>\n",
       "      <th>about_agg_lbl</th>\n",
       "      <th>gid</th>\n",
       "      <th>pages_in_domain</th>\n",
       "      <th>is_sole_page</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>firm_name</th>\n",
       "      <th>url</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">3M Innovative Properties Company</th>\n",
       "      <th>https://www.3m.com/</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://www.3m.com/3M/en_US/company-us/site-map/</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://www.3m.com/3M/en_US/company-us/about-3m/</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://www.3m.com/3M/en_US/home-improvement-us/</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://www.3m.com/3M/en_US/company-us/SDS-search/</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://www.3m.com/3M/en_US/company-us/patent/</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://www.3m.com/3M/en_US/company-us/help-center/</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ACACIA RESEARCH GROUP LLC</th>\n",
       "      <th>http://acaciaresearch.com/corporate-governance/</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://acaciaresearch.com/history/</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://acaciaresearch.com/corporate-profile/</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                     about_lbl  \\\n",
       "firm_name                        url                                                             \n",
       "3M Innovative Properties Company https://www.3m.com/                                       0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/site-map/          0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/about-3m/          1.0   \n",
       "                                 https://www.3m.com/3M/en_US/home-improvement-us/          0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/SDS-search/        0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/patent/            0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/help-cen...        0.0   \n",
       "ACACIA RESEARCH GROUP LLC        http://acaciaresearch.com/corporate-governance/           0.0   \n",
       "                                 http://acaciaresearch.com/history/                        0.0   \n",
       "                                 http://acaciaresearch.com/corporate-profile/              1.0   \n",
       "\n",
       "                                                                                     mgmt_lbl  \\\n",
       "firm_name                        url                                                            \n",
       "3M Innovative Properties Company https://www.3m.com/                                      0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/site-map/         0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/about-3m/         0.0   \n",
       "                                 https://www.3m.com/3M/en_US/home-improvement-us/         0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/SDS-search/       0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/patent/           0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/help-cen...       0.0   \n",
       "ACACIA RESEARCH GROUP LLC        http://acaciaresearch.com/corporate-governance/          1.0   \n",
       "                                 http://acaciaresearch.com/history/                       0.0   \n",
       "                                 http://acaciaresearch.com/corporate-profile/             0.0   \n",
       "\n",
       "                                                                                     partners_lbl  \\\n",
       "firm_name                        url                                                                \n",
       "3M Innovative Properties Company https://www.3m.com/                                          0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/site-map/             0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/about-3m/             0.0   \n",
       "                                 https://www.3m.com/3M/en_US/home-improvement-us/             0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/SDS-search/           0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/patent/               0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/help-cen...           0.0   \n",
       "ACACIA RESEARCH GROUP LLC        http://acaciaresearch.com/corporate-governance/              0.0   \n",
       "                                 http://acaciaresearch.com/history/                           0.0   \n",
       "                                 http://acaciaresearch.com/corporate-profile/                 0.0   \n",
       "\n",
       "                                                                                     ip_lbl  \\\n",
       "firm_name                        url                                                          \n",
       "3M Innovative Properties Company https://www.3m.com/                                    0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/site-map/       0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/about-3m/       0.0   \n",
       "                                 https://www.3m.com/3M/en_US/home-improvement-us/       0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/SDS-search/     0.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/patent/         1.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/help-cen...     0.0   \n",
       "ACACIA RESEARCH GROUP LLC        http://acaciaresearch.com/corporate-governance/        0.0   \n",
       "                                 http://acaciaresearch.com/history/                     0.0   \n",
       "                                 http://acaciaresearch.com/corporate-profile/           0.0   \n",
       "\n",
       "                                                                                     about_agg_lbl  \\\n",
       "firm_name                        url                                                                 \n",
       "3M Innovative Properties Company https://www.3m.com/                                             0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/site-map/                0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/about-3m/                1   \n",
       "                                 https://www.3m.com/3M/en_US/home-improvement-us/                0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/SDS-search/              0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/patent/                  0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/help-cen...              0   \n",
       "ACACIA RESEARCH GROUP LLC        http://acaciaresearch.com/corporate-governance/                 1   \n",
       "                                 http://acaciaresearch.com/history/                              0   \n",
       "                                 http://acaciaresearch.com/corporate-profile/                    1   \n",
       "\n",
       "                                                                                     gid  \\\n",
       "firm_name                        url                                                       \n",
       "3M Innovative Properties Company https://www.3m.com/                                   1   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/site-map/      1   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/about-3m/      1   \n",
       "                                 https://www.3m.com/3M/en_US/home-improvement-us/      1   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/SDS-search/    1   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/patent/        1   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/help-cen...    1   \n",
       "ACACIA RESEARCH GROUP LLC        http://acaciaresearch.com/corporate-governance/       4   \n",
       "                                 http://acaciaresearch.com/history/                    4   \n",
       "                                 http://acaciaresearch.com/corporate-profile/          4   \n",
       "\n",
       "                                                                                     pages_in_domain  \\\n",
       "firm_name                        url                                                                   \n",
       "3M Innovative Properties Company https://www.3m.com/                                            16.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/site-map/               16.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/about-3m/               16.0   \n",
       "                                 https://www.3m.com/3M/en_US/home-improvement-us/               16.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/SDS-search/             16.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/patent/                 16.0   \n",
       "                                 https://www.3m.com/3M/en_US/company-us/help-cen...             16.0   \n",
       "ACACIA RESEARCH GROUP LLC        http://acaciaresearch.com/corporate-governance/                 4.0   \n",
       "                                 http://acaciaresearch.com/history/                              4.0   \n",
       "                                 http://acaciaresearch.com/corporate-profile/                    4.0   \n",
       "\n",
       "                                                                                     is_sole_page  \n",
       "firm_name                        url                                                               \n",
       "3M Innovative Properties Company https://www.3m.com/                                          0.0  \n",
       "                                 https://www.3m.com/3M/en_US/company-us/site-map/             0.0  \n",
       "                                 https://www.3m.com/3M/en_US/company-us/about-3m/             0.0  \n",
       "                                 https://www.3m.com/3M/en_US/home-improvement-us/             0.0  \n",
       "                                 https://www.3m.com/3M/en_US/company-us/SDS-search/           0.0  \n",
       "                                 https://www.3m.com/3M/en_US/company-us/patent/               0.0  \n",
       "                                 https://www.3m.com/3M/en_US/company-us/help-cen...           0.0  \n",
       "ACACIA RESEARCH GROUP LLC        http://acaciaresearch.com/corporate-governance/              0.0  \n",
       "                                 http://acaciaresearch.com/history/                           0.0  \n",
       "                                 http://acaciaresearch.com/corporate-profile/                 0.0  "
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(labeled_urls[:10])\n",
    "print(about_labels[:10])\n",
    "print(mgmt_labels[:10])\n",
    "print(partner_labels[:10])\n",
    "print(ip_labels[:10])\n",
    "df_about_labeled.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load full document features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code from extract-pages-from-mongo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "def strip_firm_name (firm_name, text):\n",
    "    strip_regex = re.compile(r\"(\" + \"\\s|\".join(firm_name.split()) + \"\\s)\", re.IGNORECASE)\n",
    "    clnd_text = strip_regex.sub ('', text)\n",
    "    \n",
    "    more_regex = re.compile(r\"([A-Z]\\.?){1,} \")\n",
    "    clnd_text = more_regex.sub ('', clnd_text)\n",
    "    \n",
    "    return ' '.join(clnd_text.split(' '))\n",
    "\n",
    "# standard firm cleaning regex\n",
    "def clean_firm_name (firm):\n",
    "    firm_clnd = re.sub('(\\.|,| corporation| incorporated| llc| inc| international| gmbh| ltd)', '', firm, flags=re.IGNORECASE).rstrip()\n",
    "    return firm_clnd\n",
    "\n",
    "# remove html content\n",
    "def is_javascript (x):\n",
    "    match_string = r\"(CDATA|return\\s+true|return\\s+false|getelementbyid|function|\\w+\\(.*?\\);|\\w{2,}[\\\\.|:]+\\w{2,}|header|hover|'\\w+':\\s+'\\w+|\\\\|{|}|\\r|\\n|\\/\\/')\"\n",
    "    # capture CDATA; function declarations; function calls; word sequences separated by a period (e.g., denoting paths)\n",
    "    regex = re.findall(match_string, x) \n",
    "    # check to see if the regex finds some percentage of the words look like javascript patterns\n",
    "    if (len(regex) / float(len(x.split())) > .10):\n",
    "        return True \n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def clean_page_content (text_list):\n",
    "    # remove whatever we think is html\n",
    "    removed_html = filter(lambda x: not( bool(BeautifulSoup(x, \"html.parser\").find()) ), text_list)\n",
    "    # remove content that looks like javascript \n",
    "    removed_js = filter(lambda x: not (is_javascript(x)), removed_html)\n",
    "    # add other checks here as needed\n",
    "\n",
    "    return removed_js\n",
    "    \n",
    "# iterate through firm urls and return concatenated string\n",
    "def get_content (firm_name, urls): \n",
    "    url_to_text_dict = {}\n",
    "    for url in urls:\n",
    "        running_text = ''\n",
    "        # print ('\\tWorking on ' + url)\n",
    "        result = col.find_one( {\"url\": url} )\n",
    "        if result:\n",
    "            clnd_text = clean_page_content(result['full_text'])\n",
    "            clnd_text = '\\n'.join(clnd_text)\n",
    "            boilerpipe = None\n",
    "            \n",
    "            if 'body' in result:\n",
    "                extractor = Extractor(extractor='DefaultExtractor', html = result['body'][0])\n",
    "                lines = extractor.getText().replace(u'\\xa0', u' ').split('\\n')\n",
    "                filtered = filter(lambda x: not re.match(r'^\\s*$', x), lines)\n",
    "                boilerpipe = '\\n'.join(filtered)\n",
    "\n",
    "            # TODO fix to split().  Counting characters currently \n",
    "            if boilerpipe and (len(boilerpipe) > .5 * len(clnd_text)):\n",
    "                # print ('\\t\\tUsing boilerplate')\n",
    "                running_text += boilerpipe\n",
    "            else:\n",
    "                # print ('\\t\\tUsing clnd_text')\n",
    "                running_text += clnd_text\n",
    "            \n",
    "            running_text_clnd = strip_firm_name (firm_name, running_text)\n",
    "            url_to_text_dict[url] = running_text.split('\\n')\n",
    "        else:\n",
    "            print ('Cannot find url: ' + url)\n",
    "\n",
    "    return url_to_text_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run process_firm and write to file\n",
    "pp = pprint.PrettyPrinter()\n",
    "\n",
    "url_to_text_dict = {}\n",
    "for firm_name in firm_names: \n",
    "    # print (\"Working on \" + firm_name)\n",
    "    results = col.find ({'firm_name': firm_name})\n",
    "    if not results:\n",
    "        print (\"\\tCouldn't find any urls for firm!\")\n",
    "        continue\n",
    "    \n",
    "    urls = []\n",
    "    for result in results: \n",
    "        urls.append (result['url'][0])\n",
    "    \n",
    "    # pp.pprint (urls) \n",
    "    # print (len(urls))\n",
    "    return_dict = get_content (firm_name, urls)\n",
    "    url_to_text_dict = {**url_to_text_dict, **return_dict}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Skip to main content',\n",
      " '2Blades Foundation',\n",
      " 'The Unmet Need',\n",
      " 'How We Work',\n",
      " 'Our Team',\n",
      " 'Overview',\n",
      " 'Management & Directors',\n",
      " '2Blades Group, The Sainsbury Laboratory',\n",
      " 'Scientific Advisory Board',\n",
      " '2Blades Groups',\n",
      " 'Overview',\n",
      " 'Norwich, UK',\n",
      " 'Projects & Technology',\n",
      " 'Current Projects',\n",
      " 'Technology Patent Estate and Development',\n",
      " 'Project Grants',\n",
      " 'Potential Future Projects',\n",
      " 'News & Media',\n",
      " 'News',\n",
      " 'In the News',\n",
      " 'Newsletters',\n",
      " 'Infographics',\n",
      " 'Videos',\n",
      " 'Management & Directors',\n",
      " 'Dr. Roger Freedman',\n",
      " 'Chairman',\n",
      " '2Blades Projects',\n",
      " 'Wheat Stem Rust',\n",
      " 'Dr. Roger Freedman works for David Sainsbury, the settlor of the Gatsby '\n",
      " 'Foundation and until recently the UK Minister of Science and Innovation. Dr. '\n",
      " 'Freedman established the Gatsby Foundation’s programs in plant science and '\n",
      " 'neuroscience and continues to serve the Foundation’s much expanded plant '\n",
      " 'science program. He has been a member of the governing council of the '\n",
      " 'Sainsbury Laboratory in Norwich since its establishment and is a member of '\n",
      " 'the management board of the second Sainsbury Laboratory now established in '\n",
      " 'Cambridge. He initiated development of JICI (now PBL), the technology '\n",
      " 'licensing company based at the John Innes Centre, remaining as a '\n",
      " 'non-executive director of the company for the first ten years of its '\n",
      " 'operation. In parallel with these activities, he has also been involved in '\n",
      " 'developing the practical applications of new technologies through commercial '\n",
      " 'development. Dr. Freedman has advised venture capital funds in the US and '\n",
      " 'Europe on plant biotechnology investments and has acted as a scientific '\n",
      " 'advisor, research director, director or chairman of start up technology '\n",
      " 'companies in the UK and the US. Dr. Freedman got his B.A. at Cambridge and '\n",
      " 'received his Ph.D. for research in molecular genetics at the Laboratory of '\n",
      " 'Molecular Biology, before getting his first introduction to America as a '\n",
      " 'postdoc at the Stanford Department of Biochemistry.',\n",
      " 'Dr. Diana Horvath',\n",
      " 'President, Director',\n",
      " '847-425-1277',\n",
      " '2Blades Projects',\n",
      " 'Late Blight of Potato in Africa',\n",
      " 'Tomato Bacterial Diseases',\n",
      " 'Dr. Diana Horvath’s interests are in the delivery of seed with improved '\n",
      " 'disease resistance, particularly for smallholder farmers.\\xa0 Before joining '\n",
      " 'Roger Freedman to set up 2Blades, she\\xa0served as Science Director at ATP '\n",
      " 'Capital, a New York venture capital firm that invested in companies '\n",
      " 'developing agricultural biotechnologies. Focusing on technology and '\n",
      " 'intellectual property, she helped to build and manage a portfolio of '\n",
      " 'companies involved in forestry biotechnology, transcriptome and proteome '\n",
      " 'analysis, small molecule discovery, and embryonic stem cells.\\xa0Prior to '\n",
      " 'that, Dr. Horvath conducted research on the\\xa0molecular mechanisms of plant '\n",
      " 'disease resistance as a National Science Foundation fellow and Zeneca Plant '\n",
      " 'Science fellow at the Rockefeller University.\\xa0\\xa0Dr. Horvath received '\n",
      " 'her\\xa0Ph.D. in biochemistry and molecular biology from Northwestern '\n",
      " 'University.\\xa0 She spent a year conducting research on\\xa0maize and millet '\n",
      " 'production in southern Africa\\xa0at the Department of Agricultural Research '\n",
      " 'in Botswana, after graduating\\xa0from Tufts University with\\xa0a BS in '\n",
      " 'Biology.',\n",
      " 'Dr. T. Lynne Reuber',\n",
      " 'Program Director',\n",
      " 'Related Links',\n",
      " '2Blades Masterclass',\n",
      " '2Blades Projects',\n",
      " 'Citrus Canker',\n",
      " 'Corn Stalk and Ear Rots',\n",
      " 'Resistance Gene Diversity',\n",
      " 'Wheat Stem Rust',\n",
      " 'Dr. Lynne Reuber’s goal is to translate the recent revolutionary academic '\n",
      " 'discoveries in plant pathology to disease resistant crops to the field.\\xa0 '\n",
      " 'Prior to joining 2Blades, Dr. Reuber was a Director of Research at Mendel '\n",
      " 'Biotechnology, Inc., where she managed an inter-company collaboration to '\n",
      " 'determine the mode of action of transcription factors providing yield and '\n",
      " 'stress tolerance traits, led new trait discovery programs in the areas of '\n",
      " 'yield, resource use efficiency, and disease resistance, and managed '\n",
      " 'technology licensing relationships.\\xa0 Dr. Reuber’s scientific background '\n",
      " 'spans the interactions of plants with both disease causing and beneficial '\n",
      " 'microbes. She conducted research as an ACS postdoctoral fellow at Harvard '\n",
      " 'Medical School on the defense reactions of plants to bacterial and fungal '\n",
      " 'pathogens, after graduate work as a National Science Foundation fellow '\n",
      " 'investigating symbiotic plant-microbe interactions.\\xa0 She holds a Ph.D. in '\n",
      " 'Biology from Massachusetts Institute of Technology and a B.A. in Biology and '\n",
      " 'Chemistry from Drury University. Dr. Reuber joined 2Blades in 2014.',\n",
      " 'Dr. Dieuwertje van der Does',\n",
      " 'Group Leader, sub-Saharan Africa projects',\n",
      " 'Related Links',\n",
      " 'BecA-ILRI Hub',\n",
      " 'Dr. van der Does is interested in advancing\\xa0research on disease resistant '\n",
      " 'crops for sub-Saharan Africa.\\xa0 She spent two years at the BecA-ILRI Hub '\n",
      " 'in Nairobi, Kenya, on projects on potato, soybean, and tomato. Before '\n",
      " 'joining 2Blades, Dr. van der Does worked as a post-doctoral researcher at '\n",
      " 'The Sainsbury Laboratory in Norwich, UK, working on danger perception in '\n",
      " 'plants and was supported by a long-term fellowship from the European '\n",
      " 'Molecular Biology Organization (EMBO). Dr. van der Does received her PhD in '\n",
      " 'plant pathology from Utrecht University, the Netherlands, and received her '\n",
      " 'B.Sc. and M.Sc. in biology from the University of Amsterdam, the '\n",
      " 'Netherlands.',\n",
      " 'Dr. Peter van Esse',\n",
      " 'Group Leader, The Sainsbury Laboratory, Norwich',\n",
      " 'Related Links',\n",
      " 'Dr. van Esse, The Sainsbury Laboratory',\n",
      " '2Blades Group',\n",
      " '2Blades Group Blog',\n",
      " 'Masterclass',\n",
      " 'Video: Interview with Peter',\n",
      " '2Blades Projects',\n",
      " 'Asian Soybean Rust',\n",
      " 'Corn Stalk and Ear Rots',\n",
      " 'Dr. Peter van Esse is the Principal Investigator of the ',\n",
      " '2Blades Group',\n",
      " ' in the Sainsbury Laboratory (TSL) in Norwich. Before joining 2Blades, Dr. '\n",
      " 'van Esse was awarded a personal VENI grant by The Netherlands Organization '\n",
      " 'for Scientific Research (NWO). This allowed him to build on his graduate and '\n",
      " 'postdoctoral research on fungal effector proteins of major pathogens of '\n",
      " 'vegetable crops at the Wageningen University in the Laboratory of '\n",
      " 'Phytopathology. Dr. van Esse received both his M.Sc. in Biotechnology and '\n",
      " 'his Ph.D. in plant pathology from the Wageningen University and Research '\n",
      " 'Centre, the Netherlands.',\n",
      " 'Susan Nycum',\n",
      " 'Director',\n",
      " 'Susan Nycum provides mediation and arbitration services in domestic and '\n",
      " 'international IP and technology related disputes.\\xa0 Formerly, she was an '\n",
      " 'international partner of Baker & McKenzie where she was chair of the firm’s '\n",
      " 'North America Intellectual Property (IP) and Information Technology (IT) '\n",
      " 'practice group, a coordinator of the firm’s global IT practice group and a '\n",
      " 'member of the firm’s governing bodies for North America and for Asia '\n",
      " 'Pacific. Ms. Nycum was a member of the National Science Foundation advisory '\n",
      " 'board that approved funding for the Internet.\\xa0 She has been an advisor to '\n",
      " 'the governments of Brazil, Canada, Finland, Germany, India, Israel, '\n",
      " 'Singapore, Thailand and the United States for IT and IP policy. Susan Nycum '\n",
      " 'is the former Chair of the ABA section of Science and Technology and former '\n",
      " 'President of the Computer Law Association. She was the Chair of the '\n",
      " 'International Bar Association committee on software protection and committee '\n",
      " 'on computer crime. She was twice the ABA appointee to the National '\n",
      " 'Conference of Lawyers and Scientists.\\xa0 She is a member of the IP panel, '\n",
      " 'the Commercial panel and the Large Complex Case panel of the American '\n",
      " 'Arbitration Association; member of the CPR Technology panel and the '\n",
      " 'Electronic Discovery panel; fellow of the College of Commercial Arbitrators, '\n",
      " 'fellow of the American Bar Foundation, fellow of the ACM, fellow of the '\n",
      " 'College of Law Practice Management; member of the California Academy of '\n",
      " 'Distinguished Neutrals, Board member of the Mediation Society and several '\n",
      " 'other international mediation organizations.\\xa0 She is listed in the Best '\n",
      " 'Lawyers of America and the Northern California Super Lawyers.',\n",
      " 'Dr. Eric Ward',\n",
      " 'Director',\n",
      " 'Related Links',\n",
      " 'AgBiome',\n",
      " 'Dr. Eric Ward is the co-CEO of AgBiome. \\xa0Between 2007 and 2012, he served '\n",
      " 'as the President of the 2Blades Foundation and as Group Leader of the '\n",
      " '2Blades Group at the Sainsbury Laboratory. \\xa0Prior to that he served as '\n",
      " 'CEO of Cropsolution, Inc., a crop protection chemical discovery company, and '\n",
      " 'before that he was Co-President of Novartis (now Syngenta) Agribusiness '\n",
      " 'Biotechnology Research, where he was responsible for a staff of 270, '\n",
      " 'including researchers and all administrative functions, including finance, '\n",
      " 'patents, business development, public affairs, human resources, and '\n",
      " 'facilities.\\xa0 Simultaneously, Dr. Ward was head of target discovery for '\n",
      " 'Novartis Crop Protection AG, where he implemented a fully integrated '\n",
      " 'agricultural chemical lead discovery program based on proprietary molecular '\n",
      " 'targets.\\xa0 This program relied on extensive interactions with biotech '\n",
      " 'firms and academic labs.\\xa0 Prior to that, he was a Research Director for '\n",
      " 'the Novartis herbicide business unit, during which time his team invented '\n",
      " 'Acuron™ herbicide tolerance technology, developed corn and sugar beet '\n",
      " 'varieties engineered with the Acuron™ gene, and built the patent strategy to '\n",
      " 'protect the technology.\\xa0 In 1994-5, he worked in Basel, Switzerland as a '\n",
      " 'project leader for Ciba Crop Protection in the Weed Control business '\n",
      " 'unit.\\xa0 Dr. Ward began his career in 1988 with Ciba-Geigy as a '\n",
      " 'postdoctoral associate, during which time he pioneered methods for cloning '\n",
      " 'of large DNA fragments from plants.\\xa0 He received his Ph.D. in plant '\n",
      " 'biology from Washington University in St. Louis in 1988, where he was a '\n",
      " 'graduate fellow of the National Science Foundation.\\xa0 He received his B.S '\n",
      " 'in biology magna cum laude from Duke University in 1982.\\xa0',\n",
      " '2Blades Group, The Sainsbury Laboratory',\n",
      " 'Dr. Peter van Esse',\n",
      " 'Group Leader',\n",
      " 'Related Links',\n",
      " 'Video: Interview with Peter',\n",
      " '2Blades Projects',\n",
      " 'Asian Soybean Rust',\n",
      " 'Corn Stalk and Ear Rots',\n",
      " 'Dr. Peter van Esse is the Principal Investigator of the ',\n",
      " '2Blades Group',\n",
      " ' in the Sainsbury Laboratory (TSL) in Norwich. Before joining 2Blades, Dr. '\n",
      " 'van Esse was awarded a personal VENI grant by The Netherlands Organization '\n",
      " 'for Scientific Research (NWO). This allowed him to build on his graduate and '\n",
      " 'postdoctoral research on fungal effector proteins of major pathogens of '\n",
      " 'vegetable crops at the Wageningen University in the Laboratory of '\n",
      " 'Phytopathology. Dr. van Esse received both his M.Sc. in Biotechnology and '\n",
      " 'his Ph.D. in plant pathology from the Wageningen University and Research '\n",
      " 'Centre, the Netherlands.',\n",
      " 'Dr. Carolina Grandellis',\n",
      " 'Research Associate',\n",
      " '2Blades Projects',\n",
      " 'Corn Stalk and Ear Rots',\n",
      " 'Dr. Carolina Grandellis joined the 2Blades Group in 2017. Her work focuses '\n",
      " 'on the discovery of novel plant immune receptors that confer durable genetic '\n",
      " 'resistance to devastating plant diseases. Prior to taking up her position, '\n",
      " 'Dr. Grandellis worked as a postdoctoral fellow at the',\n",
      " 'Institute of Molecular and',\n",
      " 'Cellular\\xa0Biology of',\n",
      " 'Rosario in Argentina',\n",
      " 'studying plant-pathogen interactions between citrus and\\xa0',\n",
      " 'Xanthomonas',\n",
      " 'citri',\n",
      " 'sp.',\n",
      " 'citri',\n",
      " ', the pathogen that causes citrus canker. She also spent three years '\n",
      " 'working\\xa0at the',\n",
      " 'Institute',\n",
      " 'of',\n",
      " 'Agrobiotechnology',\n",
      " 'of',\n",
      " 'Rosario',\n",
      " 'INDEAR in Argentina on\\xa0plant transformation. Dr. Grandellis\\xa0carried '\n",
      " 'out her Ph.D at the Genetic Engineering and Molecular Biology Institute '\n",
      " 'INGEBI in Argentina studying the role of calcium dependent protein kinases '\n",
      " 'from potato.',\n",
      " 'Ruud Grootens',\n",
      " 'Laboratory Technician',\n",
      " 'Mr. Grootens graduated with a BSc Diploma from HAN University of Applied '\n",
      " 'Science in August 2018. During his studies, he conducted an internship at '\n",
      " 'Radboud University (Nijmegen) working with ',\n",
      " 'Solanum dulcamara ',\n",
      " 'accessions to investigate the genetic background of genes involved in '\n",
      " 'glycoalkaloid biosynthesis, and their involvement in slug and snail '\n",
      " 'deterrence. Mr. Grootens also spent six months at the James Hutton Institute '\n",
      " '(Dundee, Scotland) in Dr. Edgar Huitema’s Group, using CRISPR-Cas9 mediated '\n",
      " 'gene knockout in ',\n",
      " 'Phytophthora capsici',\n",
      " '. Prior to his undergraduate studies, Mr. Grootens conducted a 6-month '\n",
      " 'internship at Wageningen University with Professor Gert Kema where he '\n",
      " 'screened ',\n",
      " 'Zymoseptoria tritici',\n",
      " ' isolates over a panel of durum wheat cultivars and identified a gene '\n",
      " 'conferring resistance to this disease. Mr. Grootens has also investigated '\n",
      " 'protein extraction from spent yeast biomass from bioreactors, and conducted '\n",
      " 'quantitative PCR analysis of aquatic pathogens at the National Institute for '\n",
      " 'Health and Welfare in Kuopio, Finland.',\n",
      " 'Dr. Yogesh Gupta',\n",
      " 'Research Associate',\n",
      " '2Blades Projects',\n",
      " 'Asian Soybean Rust',\n",
      " 'Dr. Yogesh Gupta joined the\\xa0',\n",
      " '2Blades Group',\n",
      " '\\xa0at TSL Norwich in November, 2014, to study the soybean rust pathogen, ',\n",
      " 'Phakopsora pachyrhizi',\n",
      " ', which infects legumes and cause severe yield losses to soybean production. '\n",
      " 'Before joining 2Blades, Dr. Gupta was awarded the Halpin rice blast Ph.D. '\n",
      " 'scholarship to work with Prof. Nick Talbot’s group at the University of '\n",
      " 'Exeter, UK. During his graduate studies, Dr. Gupta defined the role of the '\n",
      " 'exocyst complex in protein secretion during appressorium-mediated host '\n",
      " 'invasion by the rice blast fungus, ',\n",
      " 'Magnaporthe oryzae',\n",
      " '. Prior to this, he worked at NRCPB, New Delhi as a Senior Research Fellow '\n",
      " 'on allele mining of rice blast resistance genes from land races and wild '\n",
      " 'relatives of rice. Dr. Gupta received his M.Sc. in Agricultural '\n",
      " 'Biotechnology from the HPKV Agricultural University, Palampur, India, where '\n",
      " 'he used marker-assisted selection for pyramiding rice blast resistance genes '\n",
      " 'in susceptible rice cultivar.',\n",
      " 'Dr. Nadine Ilk',\n",
      " 'Research Associate',\n",
      " '2Blades Projects',\n",
      " 'Asian Soybean Rust',\n",
      " 'In 2014, Dr. Ilk joined the ',\n",
      " '2Blades Group',\n",
      " ' at TSL to study the genetics of disease resistance against ',\n",
      " 'Phakopsora pachyrhizi',\n",
      " ' in legumes.\\xa0 Prior to her position at 2Blades, she conducted '\n",
      " 'postdoctoral research on the genetics of developmental processes in response '\n",
      " 'to abiotic stress factors at the Max-Planck Institute for Plant Breeding '\n",
      " 'Research in Cologne (MPIPZ). Dr. Ilk was awarded a Ph.D. fellowship of the '\n",
      " 'International Max-Planck Research School and graduated ',\n",
      " 'magna cum laude',\n",
      " ' in studying the natural variation of plant performance traits in '\n",
      " 'Arabidopsis in the Department of plant breeding and genetics of Maarten '\n",
      " 'Koornneef at the MPIPZ. Before starting her doctoral studies, Dr. Ilk was a '\n",
      " 'guest lecturer at the Rajamangala University of Bangkok, Thailand. She '\n",
      " 'received her M.Sc. in biology at the Ruhr-University of Bochum, Germany. Dr. '\n",
      " 'Ilk is currently a participant in the 2Blades master class in “Preparing the '\n",
      " 'next generation of leaders and entrepreneurs in Ag-Biotech.”',\n",
      " 'Dr. Cintia Kawashima',\n",
      " 'Research Associate',\n",
      " '2Blades Projects',\n",
      " 'Asian Soybean Rust',\n",
      " 'Dr. Cintia Kawashima joined the\\xa0',\n",
      " '2Blades Group',\n",
      " '\\xa0to\\xa0develop durable race independent resistance against Asian Soybean '\n",
      " 'Rust, a devastating disease caused by the biotrophic fungal pathogen, ',\n",
      " 'Phakopsora pachyrhizi',\n",
      " '. Prior to taking up her current position in The Sainsbury Laboratory, Dr. '\n",
      " 'Kawashima worked as a postdoctoral fellow in the group of professor Tamas '\n",
      " 'Dalmay, working on the role of miRNAs in the regulation of sulphate '\n",
      " 'assimilation in Arabidopsis at the University of East Anglia in Norwich. '\n",
      " 'Prior to this research, she worked for 10 months as an academic visitor on '\n",
      " 'polyamine biosynthesis in Arabidopsis at the Institute of Food Research in '\n",
      " 'Norwich. Before moving to the UK, Dr. Kawashima was awarded a Mombukagakusho '\n",
      " 'scholarship and carried out her Ph.D on sulphur metabolism in Arabidopsis in '\n",
      " 'the group of professor Kazuki Saito at Chiba University, Japan. Dr. '\n",
      " 'Kawashima graduated in Biological Sciences from the North Fluminense State '\n",
      " 'University, Brazil and received her M.Sc. in plant biotechnology from the '\n",
      " 'same university.',\n",
      " 'Lesley Milnes',\n",
      " 'Laboratory Technician',\n",
      " '2Blades Projects',\n",
      " 'Corn Stalk and Ear Rots',\n",
      " 'Lesley Milnes joined the 2Blades Group in 2017, working with Dr. Freddy '\n",
      " 'Boutrot as a laboratory technician on research into genetic resistances to '\n",
      " 'corn fungal diseases. In 2015, Ms. Milnes\\xa0started a nine month placement '\n",
      " 'working with Agriculture & Horticulture Development Board (AHDB), '\n",
      " 'researching possible biocontrol solutions against fungal pathogens infecting '\n",
      " 'stored potatoes. Ms. Milnes\\xa0received a B.Sc. (Hons) in Plant Science from '\n",
      " 'the University of Nottingham in 2015.',\n",
      " 'Dr. Marion Navarri',\n",
      " 'Research Associate',\n",
      " '2Blades Projects',\n",
      " 'Corn Stalk and Ear Rots',\n",
      " 'Dr Marion Navarri joined the\\xa02Blades Group to work on a collaborative '\n",
      " 'project to identify genes that\\xa0confer resistance to plant diseases of '\n",
      " 'economic importance. Dr. Navarri is interested in understanding the '\n",
      " 'eliciting activities of fungal pathogens. During\\xa0her PhD\\xa0she studied '\n",
      " 'deep-sea fungi producing antimicrobial compounds against pathogens of '\n",
      " 'clinical interest\\xa0in Dr. Gaetan Burgaud’s group, at the University of '\n",
      " 'Bretagne Loire, France as part of the European project MaCuMBA (Marine '\n",
      " 'microorganisms: Cultivation methods for improving their Biotechnological '\n",
      " 'Applications). Dr. Navarri received her M.Sc. in Fundamental and Applied '\n",
      " 'Microbiology from the University of Brest and also graduated from the ESIAB '\n",
      " 'School of Engineering in Food, specializing in food microbiology and safety.',\n",
      " 'Dr. Brian Rigney',\n",
      " 'Research Associate',\n",
      " '2Blades Projects',\n",
      " 'Asian Soybean Rust',\n",
      " 'Dr. Rigney joined the 2Blades Group in 2017 to study the genetics of disease '\n",
      " 'resistance against ',\n",
      " 'Phakopsora pachyrhiz',\n",
      " 'i in diverse legume species. \\xa0Prior to his position at 2Blades, he '\n",
      " 'completed his Ph.D. in potato genetics with Dr. Dan Milbourne’s lab at '\n",
      " 'Teagasc Oak Park, Ireland. During his Ph.D., he mapped potato nematode '\n",
      " 'resistance genes and validated the efficacy of molecular markers in a '\n",
      " 'commercial breeding program to improve resistance levels towards nematodes, '\n",
      " 'through pyramiding of different resistance sources. \\xa0Dr. Rigney also '\n",
      " 'holds an\\xa0M. Res in Crop Science from The James Hutton Institute, formally '\n",
      " 'known as the Scottish Crop Research Institute (SCRI), in partnership with '\n",
      " 'The University of Dundee.',\n",
      " 'Dr. Kathrin Thor',\n",
      " 'Research Associate',\n",
      " '2Blades Projects',\n",
      " 'Asian Soybean Rust',\n",
      " 'Dr Kathrin Thor joined the 2Blades group at TSL in Norwich in July 2018 to '\n",
      " 'investigate the interaction between ',\n",
      " 'Phakopsora pachyrhizi',\n",
      " ' and soybean and other legume species. Before joining 2Blades, she worked in '\n",
      " 'the group of Cyril Zipfel at TSL to identify new players of the plant immune '\n",
      " 'signaling pathway. Prior to this she was a senior postdoc at the plant '\n",
      " 'nutrition group at the University of Halle-Wittenberg in Germany. In both '\n",
      " 'positions her main interest lay in deciphering calcium signaling components '\n",
      " 'in plant stress responses. She earned her Ph.D. at the University of Bern in '\n",
      " 'Switzerland working on peptide transporters in Arabidopsis and their role in '\n",
      " 'nitrogen uptake and allocation and received her diploma from the University '\n",
      " 'of Hohenheim in Germany.',\n",
      " 'Scientific Advisory Board',\n",
      " 'Professor Sir David Baulcombe',\n",
      " 'University of Cambridge',\n",
      " 'Related Links',\n",
      " \"Prof. Baulcombe's website\",\n",
      " 'Professor Sir David Baulcombe is a Royal Society Research Professor and the '\n",
      " 'Regius Professor of Botany in the Plant Sciences Department of the '\n",
      " 'University of Cambridge. He was a student in Botany at Leeds (BSc) and '\n",
      " 'Edinburgh (PhD) Universities. After periods in Montreal, the University of '\n",
      " 'Georgia and the\\xa0Cambridge\\xa0Plant Breeding Institute, he spent 20 years '\n",
      " 'at the Sainsbury Laboratory, Norwich. He joined Cambridge University in '\n",
      " '2007.\\xa0 Sir David is a Fellow of the Royal Society and a foreign associate '\n",
      " 'member of the US National Academy of Sciences. His awards include the 2006 '\n",
      " 'Royal Medal of the Royal Society, the 2008 Lasker Award for basic biomedical '\n",
      " 'sciences,\\xa0the Wolf Prize for Agriculture in 2010 and the 2012 Balzan '\n",
      " 'Prize. He was knighted in June 2009.\\xa0 Sir David’s interests center on '\n",
      " 'gene silencing and epigenetics – the science of how nurture can influence '\n",
      " 'nature. These topics link to disease resistance in plants and understanding '\n",
      " 'of hybrids including hybrid crops. He is also interested in the application '\n",
      " 'of science to develop sustainable agriculture.\\xa0',\n",
      " 'Professor James Carrington',\n",
      " 'Donald Danforth Plant Science Center',\n",
      " 'Related Links',\n",
      " 'Donald Danforth Plant Science Center',\n",
      " 'Prof. James C. Carrington joined the Donald Danforth Plant Science Center as '\n",
      " 'president in 2011.\\xa0 Previously he served as the Director of the Center '\n",
      " 'for Genome Research and Biocomputing, the Stewart Professor for Gene '\n",
      " 'Research, and Distinguished Professor of Botany and Plant Pathology\\xa0at '\n",
      " 'Oregon State University.\\xa0 He received his doctorate in Plant Pathology '\n",
      " 'from the University of California, Berkeley, and began his prominent career '\n",
      " 'as a professor in the Department of Biology at Texas A&M University, where '\n",
      " 'he stayed for nine years. Dr. Carrington also served on the faculty at '\n",
      " 'Washington State University before his tenure at OSU. His awards include the '\n",
      " 'Presidential Young Investigator Award from the National Science Foundation, '\n",
      " 'the Ruth Allen Award from the American Society for Phytopathology and the '\n",
      " 'Humboldt Research Award from the Alexander von Humboldt Foundation. He has '\n",
      " 'been elected a Member of the National Academies of Science and a Fellow of '\n",
      " 'the American Academy of Microbiology, the American Phytopathological '\n",
      " 'Society, and the American Association for the Advancement of Science.',\n",
      " 'Professor Jeff Dangl',\n",
      " 'University of North Carolina, Chapel Hill',\n",
      " 'Related Links',\n",
      " \"Prof. Dangl's website\",\n",
      " 'Prof. Jeff Dangl is an HHMI Investigator. \\xa0He is also the John N. Couch '\n",
      " 'Professor of Biology and a member of the UNC Curriculum in Genetics. \\xa0He '\n",
      " 'received dual Bachelor’s degrees in Biology and English (Modern Literature) '\n",
      " 'form Stanford University in 1981. \\xa0He received his Ph.D. in 1986 for work '\n",
      " 'concerning structure-function relationships among of chimaeric monoclonal '\n",
      " 'antibodies from the Genetics Department of the Stanford Medical School. \\xa0'\n",
      " 'In 1986, Dr. Dangl was awarded an NSF Plant Biology Fellowship to pursue '\n",
      " 'post-doctoral research at the Max Planck Institute of Plant Breeding in '\n",
      " 'Cologne, Germany, in the department of Prof. Klaus Hahlbrock. \\xa0In 1989, '\n",
      " 'he began his own group at the Max Delbrück Laboratory, also in Cologne. \\xa0'\n",
      " 'In 1995, the Dangl lab moved to the University of North Carolina at Chapel '\n",
      " 'Hill. \\xa0The Dangl lab has contributed significantly to the use of '\n",
      " 'Arabidopsis genetics as a tool to analyze plant-pathogen interactions. \\xa0'\n",
      " 'Dr. Dangl is an elected member of the US National Academy of Sciences (2007) '\n",
      " 'and the German Academy of Sciences (“Die Leopoldina”, 2003). \\xa0He is a '\n",
      " 'past member of the National Research Council’s Board of Life Sciences\\xa0 '\n",
      " 'and is a past member of the North American Arabidopsis Steering Committee '\n",
      " '(elected) and the NSF ‘Eukaryotic Genetics’ and NIH ‘Genetics, Variation and '\n",
      " 'Evolution’ grant panels. He currently a member of the reviewing editorial '\n",
      " 'boards of Science, Cell, and PLoS Biology and served as co-Editor in Chief '\n",
      " 'of Current Opinions in Plant Biology. Research in the Dangl lab is funded by '\n",
      " 'HHMI, the Gordon and Betty Moore Foundation, NIH, NSF, and DOE.',\n",
      " 'Dr. Jeff Ellis',\n",
      " 'CSIRO Plant Industry',\n",
      " 'Related Links',\n",
      " 'CSIRO',\n",
      " 'Dr. Jeff Ellis leads a research group in CSIRO Plant Industry, Australia '\n",
      " 'that was among the first to clone and characterize plant disease resistance '\n",
      " 'genes and describe their products as a new class of nucleotide binding '\n",
      " 'site-leucine rich repeat proteins. \\xa0His laboratory developed the '\n",
      " 'Activator tagging system for cloning of the L6 and M rust resistance genes '\n",
      " 'in flax in 1994 and 1997. \\xa0L6 was one of the first cloned R genes. Since '\n",
      " 'this time, the group has isolated a number of further genes that control '\n",
      " 'other rust resistance specificities and have investigated the molecular '\n",
      " 'basis of gene-for-gene resistance specificity through the study of chimeric '\n",
      " 'L alleles in transgenic flax, showing that the leucine-rich repeat region is '\n",
      " 'the main player. \\xa0Dr. Ellis and his research team remain one of the '\n",
      " 'leading laboratories in plant disease resistance research and this team has '\n",
      " 'recently isolated and characterized the flax rust avirulence genes that '\n",
      " 'correspond to the flax L5, L6 L7 and M resistance genes and using yeast two '\n",
      " 'hybrid analysis, demonstrated that there is direct interaction between L6 '\n",
      " 'and AvrL567 proteins. \\xa0In addition to working with the model flax-flax '\n",
      " 'rust system and the fundamental basis of rust resistance, the Ellis group '\n",
      " 'has a strong grains industry focus on the delivery of improved rust '\n",
      " 'resistance in wheat through increased efficiency of resistance breeding '\n",
      " 'based on gene cloning and transfer and DNA marker technology.\\xa0 Dr. '\n",
      " 'Ellis\\xa0was elected Fellow of the Royal Society in May, 2009.',\n",
      " 'Professor Jonathan D. G. Jones',\n",
      " 'The Sainsbury Laboratory, Norwich',\n",
      " 'Related Links',\n",
      " \"Prof. Jones' website\",\n",
      " 'Prof. Jonathan D G Jones FRS is a leading researcher in plant/microbe '\n",
      " 'interactions. \\xa0He graduated in Botany from the Cambridge Natural Sciences '\n",
      " 'tripos (1976) and completed his Ph.D. on cereal chromosomes supervised by '\n",
      " 'Dick Flavell at the Plant Breeding Institute, Trumpington, in 1980.\\xa0 Dr. '\n",
      " 'Jones was a postdoctoral fellow with Fred Ausubel at Harvard University in '\n",
      " '1981 and 1982, working on symbiotic nitrogen fixation.\\xa0 From 1983-1988, '\n",
      " 'he worked in the private sector at a startup agbiotech company (Advanced '\n",
      " 'Genetic Sciences, Oakland, California) founded to exploit new developments '\n",
      " 'in molecular biology for crop improvement. \\xa0In 1988, he moved to the UK '\n",
      " 'to be one of the first recruits at The Sainsbury Laboratory, Norwich.\\xa0 He '\n",
      " 'has published landmark papers on plant resistance genes and mechanisms, and '\n",
      " 'on investigating the effector complements of the oomycete pathogens ',\n",
      " 'Hyaloperonospora arabidopsidis',\n",
      " ' and ',\n",
      " 'Albugo candida',\n",
      " ' using next-generation sequencing methods.\\xa0 He also pioneers innovative '\n",
      " 'methods to accelerate cloning and deployment of novel genes for resistance '\n",
      " 'to , the potato late blight pathogen. Dr. Jones has co-founded 2 companies; '\n",
      " 'Mendel Biotechnology, founded in 1997 to carry out genomics experiments to '\n",
      " 'discover and exploit key regulators of crop productivity, and Norfolk Plant '\n",
      " 'Sciences Ltd, to combine health promoting traits and disease resistance '\n",
      " 'traits in potato and tomato.\\xa0 Dr. Jones was elected a Professor at the '\n",
      " 'University of East Anglia in 1997, a member of EMBO in 1998, Fellow of the '\n",
      " 'Royal Society in 2003, and Foreign Associate of the US National Academy of '\n",
      " 'Sciences in 2015. \\xa0He has participated in several Royal Society working '\n",
      " 'groups on food security.',\n",
      " 'Professor Sophien Kamoun',\n",
      " 'The Sainsbury Laboratory, Norwich',\n",
      " 'Related Links',\n",
      " \"Prof. Kamoun's website\",\n",
      " 'Prof. Sophien Kamoun joined The Sainsbury Laboratory in 2007 and served as '\n",
      " 'Head of Laboratory in 2009-2014. Dr. Kamoun received his B.S. degree from '\n",
      " 'Pierre and Marie Curie University, and his Ph.D. in Genetics from the '\n",
      " 'University of California at Davis in 1991. He then was a postdoctoral fellow '\n",
      " 'at the NSF Center for Engineering Plants for Resistance Against Pathogens, '\n",
      " 'UC Davis, and at the Department of Phytopathology, Wageningen University, '\n",
      " 'Netherlands. From 1998-2007, Dr. Kamoun carried out research on oomycete '\n",
      " 'molecular genetics on the faculty at the Ohio State University, Department '\n",
      " 'of Plant Pathology, Wooster campus. At the Sainsbury Laboratory, Dr. Kamoun '\n",
      " 'continues to exploit genomics resources to improve understanding of plant '\n",
      " 'pathosystems, unravel novel processes and concepts in plant-microbe '\n",
      " 'interactions, and devise original disease management strategies based on the '\n",
      " 'gained knowledge. Throughout his career, Dr. Kamoun made a number of '\n",
      " 'significant contributions to the science of molecular plant pathology. Dr. '\n",
      " 'Kamoun pioneered genomics and molecular biology methods to reveal '\n",
      " 'fundamental insights into the biology of eukaryotic plant pathogens. He '\n",
      " 'discovered several effector families from pathogenic oomycetes and fungi, '\n",
      " 'and showed how they can suppress plant immunity. He demonstrated how '\n",
      " 'antagonistic coevolution with host plants accelerates evolution of effector '\n",
      " 'genes and emergence of new virulent pathogens. His pioneering work in plant '\n",
      " 'pathology has resulted in new approaches to control of some of the world’s '\n",
      " 'most serious crop diseases. Dr. Kamoun is a Thomson Reuters Highly Cited '\n",
      " 'Researcher (2014-2017). He received several awards, notably the American '\n",
      " 'Phytopathological Society Syngenta Award in 2003, the Ohio State University '\n",
      " 'Pomerene Teaching Award in 2004, the WE. Krauss Award in 2006, the Daiwa '\n",
      " 'Adrian Prize in 2010, and the APS Noel Keen Award in 2013. Recently, he was '\n",
      " 'recognized by award of the Kuwait Prize 2017 in recognition of his '\n",
      " 'international leadership in pathology research and development of '\n",
      " 'disease-resistant crop species. He was elected to Academia Europaea in 2011, '\n",
      " 'EMBO in 2015, and has won successive European Research Council (ERC) '\n",
      " 'Advanced Investigator grants in 2011 and 2017.',\n",
      " 'Professor Dr. Paul Schulze-Lefert',\n",
      " 'Max-Planck-Institut für Züchtungsforschung (MPIZ), Cologne',\n",
      " 'Related Links',\n",
      " \"Prof. Dr. Schulze-Lefert's website\",\n",
      " 'Prof. Dr. Paul Schulze-Lefert was trained in biochemistry and genetics at '\n",
      " 'Marburg, Freiburg, and Cologne Universities, Germany. \\xa0After a Ph.D. '\n",
      " 'thesis on cis-and trans-active factors regulating plant gene expression in '\n",
      " 'response to light, he became interested in fundamental processes controlling '\n",
      " 'plant microbe interactions. \\xa0Major research areas are the innate immune '\n",
      " 'system of plants\\xa0and functions of the plant microbiota.\\xa0 Dr. '\n",
      " 'Schulze-Lefert worked from 1989 to 1990 as postdoctoral fellow in Francesco '\n",
      " 'Salamini’s department at the MPIZ Cologne on the development of DNA marker '\n",
      " 'technologies in plant genomes. \\xa0In 1991 he started his own research group '\n",
      " 'at the RWTH Aachen with a focus on plant disease resistance mechanisms to '\n",
      " 'fungal pathogens. \\xa0From 1995 to 2000, he held a senior research position '\n",
      " 'and supervised a research team in the Sainsbury Laboratory at the John Innes '\n",
      " 'Centre, England. \\xa0Since 2000 he is head of the Department of Plant '\n",
      " 'Microbe Interactions at the Max-Planck-Institut für Züchtungsforschung '\n",
      " '(MPIZ), Cologne, and Honorary Professor at the University of Cologne since '\n",
      " '2003. \\xa0Dr. Schulze-Lefert is an elected EMBO member since 2006. \\xa0In '\n",
      " '2010 he was elected to the National Academy of Sciences, USA, and to the '\n",
      " '“Deutsche Akademie der Naturforscher Leopoldina”, Germany. \\xa0He was '\n",
      " 'elected Fellow to the American Academy of Microbiology in 2011. \\xa0Much of '\n",
      " 'his current work is dedicated to bridging traditional research areas like '\n",
      " 'genetics, biochemistry, and cell biology in the endeavor of increasing our '\n",
      " 'understanding of the molecular mechanisms that control plant microbe '\n",
      " 'interactions.\\xa0',\n",
      " 'Professor Brian J. Staskawicz',\n",
      " 'University of California, Berkeley',\n",
      " 'Related Links',\n",
      " \"Prof. Staskawicz's website\",\n",
      " 'Prof. Brian John Staskawicz, Ph.D. is the Maxine J. Elliot Professor of '\n",
      " 'Plant and Microbial Biology Department at the University of California, '\n",
      " 'Berkeley. Dr. Staskawicz received his Ph.D. from University of California, '\n",
      " 'Berkeley, M.S. from Yale University and B.A. from Bates College. \\xa0Dr. '\n",
      " 'Staskawicz is recognized as a world leading plant molecular pathologist. \\xa0'\n",
      " 'His contributions include cloning the first bacterial effector gene and '\n",
      " 'plant disease resistant genes. \\xa0His current work involves elucidation of '\n",
      " 'the molecular basis of plant innate immunity in both the model plant '\n",
      " 'Arabidopsis thaliana and the agronomically important plants. \\xa0A major '\n",
      " 'focus in the laboratory currently involves employing genomic strategies '\n",
      " 'including gene editing to engineer durable disease resistance in several '\n",
      " 'crops species. He was elected a member of National Academy of Sciences in '\n",
      " '1998 and is a Fellow of the American Phytopathological Society and American '\n",
      " 'Academy of Microbiology. \\xa0Dr. Staskawicz is the recipient of a number of '\n",
      " 'awards, including being named a Fulbright Scholar in 1991, the American '\n",
      " 'Phytopathological Society’s Ruth Allen Award in 1995, the U.S.D.A. Honors '\n",
      " 'Award in 1995, American Phytopathological Society’s Noel T. Keen Award for '\n",
      " 'Excellence in Molecular Plant Pathology in 2004 and\\xa0 received a Doctor '\n",
      " 'Honoris Causa, Wageningen University in 2013.\\xa0Dr. Staskawicz serves as an '\n",
      " 'external advisor to the Sainsbury Laboratory in Norwich. \\xa0He also serves '\n",
      " 'as an editor for P.N.A.S., Cell Host-Microbe.',\n",
      " 'Professor Cyril Zipfel',\n",
      " 'Institute of Plant & Microbial Biology, University of Zurich',\n",
      " 'Professor Zipfel holds the Chair of Molecular & Cellular Plant Physiology at '\n",
      " 'the University of Zurich (Switzerland) (since June 2018). Before that, he '\n",
      " 'was Senior Group Leader and Head of The Sainsbury Laboratory (TSL) in '\n",
      " 'Norwich (UK) where he had his group since 2007. He performed his doctoral '\n",
      " '(2001-2005) and post-doctoral research (2005-2007) in the laboratories of '\n",
      " 'Profs. Thomas Boller (Friedrich-Miescher Institute, Basel, Switzerland) and '\n",
      " 'Jonathan Jones (TSL Norwich, UK) (supported by an EMBO Long-Term '\n",
      " 'Post-Doctoral Fellowship), respectively. He is a pioneer and leader in the '\n",
      " 'field of plant innate immunity and receptor kinases. His work is focused on '\n",
      " 'understanding the molecular basis of plant innate immunity mediated by '\n",
      " 'surface-localized immune receptors, as well as the application of this '\n",
      " 'research to engineer disease resistance in crops. He was awarded competitive '\n",
      " 'European Research Council grants in 2012 and 2018, is a Highly Cited '\n",
      " 'Researcher (since 2014), and was awarded the Charles Albert Shull Award from '\n",
      " 'the American Society of Plant Biologists in 2015 and the 4th Tsuneko & Reiji '\n",
      " 'Okazaki Award from Nagoya University in 2018. He was elected to the European '\n",
      " 'Molecular Biology Organisation in 2018. He also assumes editorial duties for '\n",
      " 'the journals Science Signaling, Annual Review of Plant Biology, PLoS '\n",
      " 'Pathogens, Molecular Plant, and EMBO Reports.',\n",
      " 'Elsewhere on the site',\n",
      " 'Sign up for our newsletter',\n",
      " 'Thanks for signing up!',\n",
      " 'Subscribe',\n",
      " 'Bridging the Gap',\n",
      " 'Important crop disease resistance solutions have failed to move beyond the '\n",
      " 'lab bench.  2Blades bridges the gap between the laboratory and the field, '\n",
      " 'translating research in durable disease resistance into practical '\n",
      " 'applications.',\n",
      " 'Navigate',\n",
      " 'The Unmet Need',\n",
      " 'How We Work',\n",
      " 'Our Team',\n",
      " 'Overview',\n",
      " 'Management & Directors',\n",
      " '2Blades Group, The Sainsbury Laboratory',\n",
      " 'Scientific Advisory Board',\n",
      " '2Blades Groups',\n",
      " 'Overview',\n",
      " 'Norwich, UK',\n",
      " 'Projects & Technology',\n",
      " 'Current Projects',\n",
      " 'Technology Patent Estate and Development',\n",
      " 'Project Grants',\n",
      " 'Potential Future Projects',\n",
      " 'News & Media',\n",
      " 'News',\n",
      " 'In the News',\n",
      " 'Newsletters',\n",
      " 'Infographics',\n",
      " 'Videos',\n",
      " 'About 2Blades',\n",
      " 'Learn More',\n",
      " 'Infographics',\n",
      " 'Data Resources',\n",
      " 'Newsletters',\n",
      " 'Major crop disease outbreaks',\n",
      " 'Privacy Policy',\n",
      " 'Get involved',\n",
      " 'Contact',\n",
      " '2Blades Foundation',\n",
      " '+1 847 425 1277',\n",
      " 'Twitter',\n",
      " 'YouTube',\n",
      " '&copy2019 2Blades Foundation. All rights reserved.']\n"
     ]
    }
   ],
   "source": [
    "pp.pprint (url_to_text_dict['http://2blades.org/our-team/'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code from topic-model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pattern regex to remove unwanted words that show up in topic models\n",
    "p = re.compile(r\"(\\(\\)|''|``|\\\"|null|ul|li|ol|^\\.|^:|^/|\\\\|--|cooki|'s|corpor|busi|inc\\.|ltd|co\\.|compan|keyboard|product|technolog)\", flags=re.IGNORECASE)\n",
    "\n",
    "def encode_item(text):\n",
    "    '''\n",
    "    Remove characters with encoding problems \n",
    "    '''\n",
    "    clean = \"\"\n",
    "    for item in text.split(\" \"):\n",
    "        try:\n",
    "            clean += item.encode('ascii','ignore').decode('utf-8') + \" \"\n",
    "        except: \n",
    "            pass\n",
    "    return clean.rstrip()\n",
    "\n",
    "stopwords = nltk.corpus.stopwords.words('english') + list(set(string.punctuation))\n",
    "stemmer = SnowballStemmer(\"english\")\n",
    "\n",
    "#tokenize text for topic modelling\n",
    "def stem_tokenizer(text):\n",
    "    '''\n",
    "    Tokenize and stem an input sentence\"\n",
    "    '''\n",
    "    words_stemmed = list(filter(lambda x : not x in stopwords, [stemmer.stem(word) for word in nltk.word_tokenize(text)]))\n",
    "    return words_stemmed\n",
    "\n",
    "def clean_text (text):\n",
    "    clnd_text = list(filter(lambda x : not p.match(x), text))\n",
    "    return clnd_text\n",
    "\n",
    "def return_for_model(text):\n",
    "    '''\n",
    "    Deal with endcoding and return stemmed/tokenized text\n",
    "    '''\n",
    "    all_content = [encode_item(item) for item in text] \n",
    "    all_content_stem_token = list(map(stem_tokenizer, all_content))\n",
    "    all_content_clnd = list(map(clean_text, all_content_stem_token))\n",
    "    all_content_trimmed = [c for c in all_content_clnd if len(c) >= MIN_PARA_LEN ]\n",
    "    \n",
    "    all_content_no_dups = []\n",
    "    seen = {}\n",
    "    for trimmed in all_content_trimmed:\n",
    "        trimmed_joined = '|'.join(trimmed)\n",
    "        if trimmed_joined not in seen: \n",
    "            all_content_no_dups.append(trimmed)\n",
    "            seen[trimmed_joined] = 1\n",
    "        \n",
    "    return all_content_no_dups\n",
    "\n",
    "def process_doc (document):\n",
    "    # loop over text and add title elements to the paragraph they describe\n",
    "    joined = []\n",
    "    buf = ''\n",
    "    for i in range(len(document)): # figure out a way to chunk groups of content\n",
    "        if re.search (r'cookie|privacy|terms|GDPR', document[i], re.IGNORECASE):\n",
    "            continue\n",
    "        if len(document[i]) <= 12 or len(document[i].split()) < MIN_PARA_LEN : # maybe a menu or simple pagragraph heading? \n",
    "            continue\n",
    "        if re.search(r'(\\.|;|\\!)', document[i]): # found what appears to be a sentence\n",
    "            if buf:\n",
    "                joined.append(buf.strip())\n",
    "                buf = ''\n",
    "            joined.append(document[i].strip())\n",
    "        elif len(buf) > PHRASE_LENGTH: # the rest of this code processes free-standing text \n",
    "            joined.append(buf.strip())\n",
    "            buf = document[i]\n",
    "        else: \n",
    "            buf = buf + ' ' + document[i]\n",
    "    if buf:\n",
    "        joined.append(buf.rstrip())\n",
    "\n",
    "    # pp.pprint (joined)\n",
    "    return ' '.join (joined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "labeled_texts = []\n",
    "unlabeled_texts = []\n",
    "\n",
    "for url, text in url_to_text_dict.items():\n",
    "    # print (url)\n",
    "    # pp.pprint (text)\n",
    "\n",
    "    for k in [\"\\n\", \"\\t\", \" \"]: # remove the end of sentence tags\n",
    "        text = [item.strip(k) for item in text]\n",
    "    text = list(filter(lambda x : not x in [' ', '', '\\t'],text)) #remove the paragraphs that are just blank space\n",
    "    processed_text = process_doc (text)\n",
    "    \n",
    "    if url in labeled_urls:\n",
    "        labeled_texts.append(processed_text)\n",
    "    else:\n",
    "        unlabeled_texts.append(processed_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1004\n",
      "1004\n",
      "1004\n",
      "8658\n"
     ]
    }
   ],
   "source": [
    "# pp.pprint (texts)\n",
    "print (len(labeled_urls))\n",
    "print (len(labels))\n",
    "print (len(labeled_texts))\n",
    "print (len(unlabeled_texts))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recurrent LSTM neural net\n",
    "Code modified from https://medium.com/jatana/report-on-text-classification-using-cnn-rnn-han-f0e887214d5f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAX_SEQUENCE_LENGTH = 1000\n",
    "MAX_NB_WORDS = 20000\n",
    "EMBEDDING_DIM = 100\n",
    "VALIDATION_SPLIT = 0.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Unique Tokens 27306\n"
     ]
    }
   ],
   "source": [
    "tokenizer = Tokenizer(num_words=MAX_NB_WORDS)\n",
    "tokenizer.fit_on_texts(labeled_texts)\n",
    "sequences = tokenizer.texts_to_sequences(labeled_texts)\n",
    "\n",
    "word_index = tokenizer.word_index\n",
    "\n",
    "print('Number of Unique Tokens',len(word_index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of Data Tensor: (1004, 1000)\n",
      "Shape of Label Tensor: (1004,)\n",
      "[583 607 245 ... 363 552 910]\n",
      "(804, 1000)\n",
      "(804,)\n",
      "(200, 1000)\n",
      "(200,)\n"
     ]
    }
   ],
   "source": [
    "data = pad_sequences(sequences, maxlen=MAX_SEQUENCE_LENGTH)\n",
    "labels = np.array(about_labels)\n",
    "\n",
    "print('Shape of Data Tensor:', data.shape)\n",
    "print('Shape of Label Tensor:', labels.shape)\n",
    "\n",
    "indices = np.arange(data.shape[0])\n",
    "np.random.shuffle(indices)\n",
    "print (indices)\n",
    "data = data[indices]\n",
    "labels = labels[indices]\n",
    "nb_validation_samples = int(VALIDATION_SPLIT * data.shape[0])\n",
    "\n",
    "x_train = data[:-nb_validation_samples]\n",
    "y_train = labels[:-nb_validation_samples]\n",
    "x_val = data[-nb_validation_samples:]\n",
    "y_val = labels[-nb_validation_samples:]\n",
    "\n",
    "print(x_train.shape)\n",
    "print(y_train.shape)\n",
    "print(x_val.shape)\n",
    "print(y_val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total 400000 word vectors in Glove 6B 100d.\n"
     ]
    }
   ],
   "source": [
    "embeddings_index = {}\n",
    "f = open(GLOVE_DIR + 'glove.6B.100d.txt',encoding='utf8')\n",
    "for line in f:\n",
    "    values = line.split()\n",
    "    word = values[0]\n",
    "    coefs = np.asarray(values[1:], dtype='float32')\n",
    "    embeddings_index[word] = coefs\n",
    "f.close()\n",
    "\n",
    "print('Total %s word vectors in Glove 6B 100d.' % len(embeddings_index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [],
   "source": [
    "embedding_matrix = np.random.random((len(word_index) + 1, EMBEDDING_DIM))\n",
    "for word, i in word_index.items():\n",
    "    embedding_vector = embeddings_index.get(word)\n",
    "    if embedding_vector is not None:\n",
    "        # words not found in embedding index will be all-zeros.\n",
    "        embedding_matrix[i] = embedding_vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "embedding_layer = Embedding(len(word_index) + 1,\n",
    "                            EMBEDDING_DIM,\n",
    "                            weights=[embedding_matrix],\n",
    "                            input_length=MAX_SEQUENCE_LENGTH,\n",
    "                            trainable=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bidirectional LSTM\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "input_8 (InputLayer)         (None, 1000)              0         \n",
      "_________________________________________________________________\n",
      "embedding_2 (Embedding)      (None, 1000, 100)         2730700   \n",
      "_________________________________________________________________\n",
      "bidirectional_7 (Bidirection (None, 200)               160800    \n",
      "_________________________________________________________________\n",
      "dense_22 (Dense)             (None, 1)                 201       \n",
      "=================================================================\n",
      "Total params: 2,891,701\n",
      "Trainable params: 2,891,701\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "sequence_input = Input(shape=(MAX_SEQUENCE_LENGTH,), dtype='int32')\n",
    "embedded_sequences = embedding_layer(sequence_input)\n",
    "l_lstm = Bidirectional(LSTM(100))(embedded_sequences)\n",
    "preds = Dense(1, activation='softmax')(l_lstm)\n",
    "model = Model(sequence_input, preds)\n",
    "model.compile(loss='binary_crossentropy',\n",
    "              optimizer='rmsprop',\n",
    "              metrics=['acc'])\n",
    "\n",
    "print(\"Bidirectional LSTM\")\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "cp=ModelCheckpoint('model_rnn.hdf5',monitor='val_acc',verbose=1,save_best_only=True)\n",
    "history=model.fit(x_train, y_train, validation_data=(x_val, y_val),epochs=15, batch_size=2,callbacks=[cp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig1 = plt.figure()\n",
    "plt.plot(history.history['loss'],'r',linewidth=3.0)\n",
    "plt.plot(history.history['val_loss'],'b',linewidth=3.0)\n",
    "plt.legend(['Training loss', 'Validation Loss'],fontsize=18)\n",
    "plt.xlabel('Epochs ',fontsize=16)\n",
    "plt.ylabel('Loss',fontsize=16)\n",
    "plt.title('Loss Curves :RNN',fontsize=16)\n",
    "fig1.savefig('loss_rnn.png')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig2=plt.figure()\n",
    "plt.plot(history.history['acc'],'r',linewidth=3.0)\n",
    "plt.plot(history.history['val_acc'],'b',linewidth=3.0)\n",
    "plt.legend(['Training Accuracy', 'Validation Accuracy'],fontsize=18)\n",
    "plt.xlabel('Epochs ',fontsize=16)\n",
    "plt.ylabel('Accuracy',fontsize=16)\n",
    "plt.title('Accuracy Curves : RNN',fontsize=16)\n",
    "fig2.savefig('accuracy_rnn.png')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from PIL import Image\n",
    "display(Image.open('rnn_model.png'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
